File: includes/RhythmGameUtilities/Audio.cpp
#include <stdlib.h>

#include <cmath>

#include "Audio.h"

namespace RhythmGameUtilities
{

int **ConvertSamplesToWaveform(float *samples, int size, int width, int height)
{
    auto waveform = new int *[width];

    auto step = floor(size / width);
    auto amp = height / 2;

    for (auto x = 0; x < width; x += 1)
    {
        waveform[x] = new int[height];

        auto min = 1.0f;
        auto max = -1.0f;

        for (auto j = 0; j < step; j += 1)
        {
            auto index = (int)(x * step + j);

            auto datum = samples[index];

            if (datum < min)
            {
                min = datum;
            }

            if (datum > max)
            {
                max = datum;
            }
        }

        auto minY = (int)((1 + min) * amp);
        auto maxY = (int)((1 + max) * amp);

        for (auto y = 0; y < height; y += 1)
        {
            waveform[x][y] = y >= minY && y <= maxY ? 1 : 0;
        }
    }

    return waveform;
}

} // namespace RhythmGameUtilities

File: RhythmGameUtilities/Scripts/Audio.cs
using System;

namespace RhythmGameUtilities
{

    public static class Audio
    {

        /// <summary>
        /// Converts samples from an audio file into data used to display a waveform.
        /// </summary>
        ///
        /// <param name="samples">Array of sample data from an audio file.</param>
        /// <param name="width">Width of the waveform.</param>
        /// <param name="height">Height of the waveform.</param>
        public static int[][] ConvertSamplesToWaveform(float[] samples, int width, int height)
        {
            var waveform = new int[width][];

            var step = Math.Floor((double)samples.Length / width);
            var amp = height / 2;

            for (var x = 0; x < width; x += 1)
            {
                waveform[x] = new int[height];

                var min = 1.0f;
                var max = -1.0f;

                for (var j = 0; j < step; j += 1)
                {
                    var index = (int)(x * step + j);

                    var datum = samples[index];

                    if (datum < min)
                    {
                        min = datum;
                    }

                    if (datum > max)
                    {
                        max = datum;
                    }
                }

                var minY = (int)((1 + min) * amp);
                var maxY = (int)((1 + max) * amp);

                for (var y = 0; y < height; y += 1)
                {
                    waveform[x][y] = y >= minY && y <= maxY ? 1 : 0;
                }
            }

            return waveform;
        }

    }

}
